espnet2/asr/frontend/adapter_utils/adapters/adapter.py

```diff
@@ -0,0 +1,57 @@
+import torch
+import torch.nn as nn
+
+
+try:
+    from fairseq import utils
+    from fairseq.modules import LayerNorm
+
+    is_fairseq_available = True
+except ImportError:
+    is_fairseq_available = False
+
+
+class Adapter(nn.Module):
+    def __init__(
+        self,
+        orig_dim: int,
+        down_dim: int,
+        layer_norm: str = None,
+        activation_fn: str = "gelu",
+    ) -> None:
+        """
+        * orig_dim - original dimension.
+        * down_dim - dimension of Adapter's intermediate down-projection.
+        * layer_norm - location of LayerNorm layer ("first" or "last", default=None). If `None`, LayerNorm is not used.
+        * activation_fn - activation type (default="gelu").
+        """
+        super().__init__()
+        if not is_fairseq_available:
+            raise ImportError(
+                "`fairseq` is not available. Please install it via `pip install"
+                " fairseq` or `cd /path/to/espnet/tools && . ./activate_python.sh"
+                " && ./installers/install_fairseq.sh`."
+            )
+
+        self.down_projection = nn.Linear(orig_dim, down_dim)
+        self.up_projection = nn.Linear(down_dim, orig_dim)
+        nn.init.xavier_uniform_(self.down_projection.weight)
```

Comment:
Does this initialization really work as expected? I feel this [init](https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L517) will overwrite them.

Reply:
good catch , could you explain while  [https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L517](init) is fired when init is not , None ?
also as this change needs to be done in asr.py while will impact other users , what's your suggestion ?

Reply:
Hi, I've done a side-by-side comparison regarding the actual use of that two lines of code and the results suggest that the initialization somehow does work? My code tracing leaves me to this line below

https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tasks/asr.py#L518

and subsequent inquiry into the `initialize` function suggests that at the very end, after all initializations are done, there is this **reload all parameters** operation

https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/torch_utils/initialize.py#L99

and that method is implemented by loading from a previous saved `self.pretrained_params`

https://github.com/espnet/espnet/blob/b008ac7d58e9ced1a9f8c89cc85ee69d9e9461ab/espnet2/asr/frontend/s3prl.py#L48

Thus my inquiry leaves me with the conclusion that if we build model without that two lines of initialization, the model would proceed with default (therefore uninitialized adapter modules parameters), pertained s3prl frontend. With that two lines of initialization code, we would first construct and load the entire pretrained s3prl frontend and save to `self.pretrained_params`, thus successfully initializing the parameters.

Reply:
@simpleoier can we resolve this or do you have any suggestions to make this part of code more clear?

Reply:
Yes, it is resolved. I think @jwrh 's tracing explains all my concerns.